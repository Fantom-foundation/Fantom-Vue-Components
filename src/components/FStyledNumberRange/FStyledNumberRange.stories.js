import { withA11y } from '@storybook/addon-a11y';

import FStyledNumberRange from './FStyledNumberRange.vue';

export default {
    title: 'FStyledNumberRange',
    component: FStyledNumberRange,
    decorators: [withA11y],
};

export const Default = () => ({
    components: { FStyledNumberRange },
    template: `
        <div>
            <f-styled-number-range :value="75" />
        </div>
    `,
});

export const ClampValue = () => ({
    components: { FStyledNumberRange },
    template: `
        <div>
            <f-styled-number-range :value="parseFloat(value)" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 10,
        };
    },
});

export const MinMax = () => ({
    components: { FStyledNumberRange },
    template: `
        <div>
            <f-styled-number-range :value="value" :min="0" :max="30" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 50,
        };
    },
});

export const ConvertToPercentages = () => ({
    components: { FStyledNumberRange },
    template: `
        <div>
            <f-styled-number-range convert-to-percentages :value="value" :min="-50" :max="50" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 0,
        };
    },
});

export const Styles = () => ({
    components: { FStyledNumberRange },
    template: `
        <div>
            <f-styled-number-range :styles="styles" :value="parseFloat(value)" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            styles: [
                {
                    value: 0,
                    style: { color: 'red', fontWeight: 'bold' },
                },
                {
                    value: 40,
                    style: { color: '#4a5ef5' },
                },
                {
                    value: 80,
                    style: { color: 'green' },
                },
            ],
            value: 50,
        };
    },
});

export const Slot = () => ({
    components: { FStyledNumberRange },
    template: `
        <div>
            <f-styled-number-range :value="value">
                <template #default="{ value }"><strong class="fos-7">{{ value }}</strong><span class="co-grey-6">%</span></template>
            </f-styled-number-range>
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 0,
        };
    },
});
