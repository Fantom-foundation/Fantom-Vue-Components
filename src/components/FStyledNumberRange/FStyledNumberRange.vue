<template>
    <span class="fstylednumberrange" :style="cStyles">
        <slot :value="cValue" :percentage="percentage" :convertToPercentages="convertToPercentages">
            <template v-if="convertToPercentages"> {{ percentage }}% </template>
            <template v-else>
                {{ cValue }}
            </template>
        </slot>
    </span>
</template>

<script>
/**
 * Component for displaying a number styled according to the value from the given range
 */
import { clamp } from '../../utils/number.js';

export default {
    name: 'FStyledNumberRange',

    props: {
        /** Current value. */
        value: {
            type: [String, Number],
            default: 10,
        },
        /** Minimal value. */
        min: {
            type: Number,
            default: 0,
        },
        /** Maximal value. */
        max: {
            type: Number,
            default: 100,
        },
        /**
         * Css styles of number for given values.
         * If current value is bigger than value from `styles` array, css style from array will be used.
         * @type {{style: Object, value: number}[]}
         */
        styles: {
            type: Array,
            default() {
                return [];
            },
        },
        /** Convert value to percetnage */
        convertToPercentages: {
            type: Boolean,
            default: false,
        },
    },

    data() {
        return {
            numberColor: '',
        };
    },

    computed: {
        cValue() {
            let value = this.value;

            value = clamp(value, this.min, this.max);

            return value;
        },

        percentage() {
            return Math.round(((this.cValue - this.min) / (this.max - this.min)) * 100);
        },

        cStyles() {
            const { styles } = this;
            const { cValue } = this;
            let style = {};

            for (let i = styles.length - 1; i >= 0; i--) {
                if (cValue >= styles[i].value) {
                    style = styles[i].style;
                    break;
                }
            }

            return style;
        },
    },

    created() {
        if (this.max < this.min) {
            throw new Error("'max' prop must be bigger than 'min'");
        }
    },
};
</script>
