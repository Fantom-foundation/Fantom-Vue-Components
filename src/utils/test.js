/**
 * Utility functions for testing
 */

import { toKebabCase } from './string.js';

/**
 * @param {function} func Function to be called
 * @param {int} [n] Number of calls
 * @param {int} [interval] Time span between two calls in milliseconds
 * @return {Promise<unknown>}
 */
export function callFunctionNTimes(func, n = 3, interval = 25) {
    return new Promise(resolve => {
        let counter = n;
        const intervalId = setInterval(() => {
            if (counter-- <= 0) {
                clearInterval(intervalId);
                resolve();
            } else {
                func();
            }
        }, interval);
    });
}

/**
 * Disable console error messages.
 *
 * @param {function} func
 */
export function disableErrorMessages(func) {
    const originalConsoleError = console.error;
    console.error = () => {};

    func();

    console.error = originalConsoleError;
}

/**
 * Playground for testing v-model of a Vue component.
 *
 * @param {Vue} component
 * @param {Object} data
 * @return {{dValue}|{template: `<${string} v-model="dValue" ${string} />`, components: {}, data(): {dValue: *}, props: string[]}}
 */
export function getVModelComponent(component, data = {}) {
    const componentName = component.name;
    let attributes = [];

    Object.keys(data).forEach(key => {
        attributes.push(`:${key}="${key}"`);
    });

    return {
        components: { [componentName]: component },
        template: `<${toKebabCase(componentName)} v-model="dValue" ${attributes.join(' ')} />`,
        props: ['value'],
        data() {
            return {
                dValue: this.value,
                ...data,
            };
        },
    };
}

/**
 * @param {Object} wrapper
 */
export function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}
