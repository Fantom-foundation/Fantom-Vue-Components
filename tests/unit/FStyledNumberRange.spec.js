import { mount } from '@vue/test-utils';
import FStyledNumberRange from '@/components/FStyledNumberRange/FStyledNumberRange.vue';
import { disableErrorMessages } from '@/utils/test.js';

let wrapper = null;

function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}

function createWrapper({ propsData = { value: 10, min: 0, max: 100 }, slots = {} } = {}) {
    return mount(FStyledNumberRange, {
        propsData,
        slots,
    });
}

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('FStyledNumberRange', () => {
    it('should render correctly', () => {
        wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should throw an error if `min` is greater than `max`', () => {
        disableErrorMessages(function() {
            expect(() => {
                createWrapper({
                    propsData: {
                        min: 100,
                        max: 50,
                    },
                });
            }).toThrow();
        });
    });

    it('should clamp given value', async () => {
        wrapper = createWrapper({
            propsData: {
                value: -10,
            },
        });

        expect(wrapper.vm.cValue).toBe(0);

        await wrapper.setProps({ value: 200 });

        expect(wrapper.vm.cValue).toBe(100);
    });

    it('should convert given value to percentages if `convertToPercentages` prop is `true`', () => {
        wrapper = createWrapper({
            propsData: {
                value: 60,
                min: 20,
                max: 100,
                convertToPercentages: true,
            },
        });

        expect(wrapper.text()).toBe('50%');
    });

    it('should apply css style to the component if value is within given range', async () => {
        wrapper = createWrapper({
            propsData: {
                styles: [
                    {
                        value: 0,
                        style: { color: 'red' },
                    },
                    {
                        value: 40,
                        style: { color: 'orange' },
                    },
                    {
                        value: 80,
                        style: { color: 'green' },
                    },
                ],
            },
        });

        expect(wrapper.element.style.color).toBe('red');

        await wrapper.setProps({ value: 41 });
        expect(wrapper.element.style.color).toBe('orange');

        await wrapper.setProps({ value: 80 });
        expect(wrapper.element.style.color).toBe('green');
    });
});
