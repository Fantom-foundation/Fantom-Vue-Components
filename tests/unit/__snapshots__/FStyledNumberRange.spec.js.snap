// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FStyledNumberRange should render correctly 1`] = `
<span
  class="fstylednumberrange"
>
  
            10
        
</span>
`;
